### PR TITLE
Development

### PR DESCRIPTION
**Pull request by riccjohn**
Add tests for creating a new game with an empty board, placing markers on the board, and switching turns. Basic class implemented to pass tests.

### CODE REVIEW

File: src/tests/tic-tac-toe.test.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import Game from "../tic-tac-toe";
+
+describe("Tic-tac-toe", () => {
+  let game: Game;
+  beforeEach(() => {
```
Comment (rylnd) [inline src/tests/tic-tac-toe.test.ts]: I would typically separate `beforeEach`, `it`, etc. with newlines to make tests a little more readable:

```js
describe(() => {
  beforeEach()
  
  it()

  it()
})
```

File: src/tests/tic-tac-toe.test.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    game = new Game();
+  });
+  it("starts with an empty board", () => {
+    const board = [["", "", ""], ["", "", ""], ["", "", ""]];
+    expect(game.board).toEqual(board);
+  });
+  it("places an X on the board on X's turn", () => {
```
Comment (rylnd) [inline src/tests/tic-tac-toe.test.ts]: How do we know it's X's turn, here?

File: src/tests/tic-tac-toe.test.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    game.place(0, 0);
+    expect(game.board[0][0]).toEqual("X");
+  });
+  it("switches players every turn", () => {
```
Comment (rylnd) [inline src/tests/tic-tac-toe.test.ts]: This is sorta only testing one player's turn: if both placements were "0" the test would still pass.

File: src/tic-tac-toe.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+type Square = string;
+type Row = Square[];
+type Board = Row[];
+
+class Game {
+  public board: Board;
+  private currentPlayer: string;
+  constructor() {
+    this.currentPlayer = "X";
```
Comment (rylnd) [inline src/tic-tac-toe.ts]: I would try to make "X" and "Y" an Enum called `PlayerPiece` or something like that. Then you could initialize the board with `undefined` instead of empty strings (which you could still do now).

File: src/tic-tac-toe.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    this.board = [["", "", ""], ["", "", ""], ["", "", ""]];
```
Comment (rylnd) [inline src/tic-tac-toe.ts]: It would be interesting to see this generalized to an NxN board!

File: src/tic-tac-toe.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  }
+
+  switchPlayers() {
+    if (this.currentPlayer === "X") this.currentPlayer = "O";
+    else this.currentPlayer = "X";
+  }
+
+  place(x: number, y: number) {
+    this.board[x][y] = this.currentPlayer;
```
Comment (rylnd) [inline src/tic-tac-toe.ts]: I like that this implementation could potentially allow for more than two players!

File: src/tests/tic-tac-toe.test.ts (hunk 1 above)
Comment (riccjohn) [reply src/tests/tic-tac-toe.test.ts]: Agreed! Ended up doing this after the PR.

File: src/tests/tic-tac-toe.test.ts (hunk 2 above)
Comment (riccjohn) [reply src/tests/tic-tac-toe.test.ts]: Maybe this should just be testing that we're able to mark the board at all. The next test checks that the player is switched each turn, so maybe it's best to describe this assertion as making sure we can mark the board, and let the next test deal with which player's turn it is?

File: src/tests/tic-tac-toe.test.ts (hunk 3 above)
Comment (riccjohn) [reply src/tests/tic-tac-toe.test.ts]: I can add two expectations inside this test. One that expects the first move is 'X', and the other to expect that the second move is 'O'.

**Comment by riccjohn** [comment]
I did some other work on this branch, so I'm going to commit that, and then do a second commit where I address all these changes. I'll make sure to address the branch name on the next branch I create.

File: src/tests/tic-tac-toe.test.ts (hunk 2 above)
Comment (rylnd) [reply src/tests/tic-tac-toe.test.ts]: Yeah, I always waffle on these kinds of "redundant" tests, but my preference is usually to have clear test descriptions (even if the test code itself isn't quite as clear).

Even just changing the description here to `it("starts on the X players turn")` and changing nothing else would be fine; my comment was mainly about clear intent.

File: src/tic-tac-toe.ts (hunk 5 above)
Comment (riccjohn) [reply src/tic-tac-toe.ts]: Added ability to dictate board size when a new game is created. Defaults to a 3x3 board.

File: src/tic-tac-toe.ts (hunk 4 above)
Comment (riccjohn) [reply src/tic-tac-toe.ts]: Done!

File: src/tic-tac-toe.ts [hunk 7]
```diff
@@ -69,12 +69,13 @@ class Game {
...
-      this.checkColumnsForWin() ||
-      this.checkDiagonalsForWin();
+    const winCondition = [
+      this.checkRowsForWin(),
+      this.checkColumnsForWin(),
+      this.checkDiagonalsForWin()
+    ];
 
-    if (winCondition) {
+    if (winCondition.some(winnerCheck => winnerCheck === true)) {
```
Comment (rylnd) [inline src/tic-tac-toe.ts]: Since the function passed to `some` expects a boolean to be returned, and the values we're iterating _are_ booleans, this function could just be the [identity function](https://en.wikipedia.org/wiki/Identity_function): 
`winConditions.some(condition => condition)`

Comment (riccjohn) [reply src/tic-tac-toe.ts]: Great! I wasn't positive if that was the case, but knew it was weird to write `=== true`. Good to know
